datafusion/optimizer/src/optimize_projections.rs

```diff
@@ -162,14 +163,37 @@ fn optimize_projections(
                 .map(|input| ((0..input.schema().fields().len()).collect_vec(), false))
                 .collect::<Vec<_>>()
         }
+        LogicalPlan::Extension(extension) => {
+            let children = extension.node.inputs();
+            if children.len() != 1 {
+                // TODO: Add support for `LogicalPlan::Extension` with multi children.
+                return Ok(None);
+            }
+            // has single child
+            let exprs = plan.expressions();
+            let child = children[0];
+            let node_schema = extension.node.schema();
+            let child_schema = child.schema();
+            if let Some(parent_required_indices_mapped) =
```

Comment:
I think the information that is needed is "given the parent only needs some of the output columns of this UserDefinedPlan, can it avoid requiring additional input columns"

Given the UserDefinedPlan is basically a black box and the optimizer doesn't know how it works, I think the only way to get this information would be to add some sort of API to UserDefined node

I suggest for this PR, only push down information about used expressions (don't try and incorporate a projection from the parent) and we can handle the parent projection information in a follow on PR


I am thinking of an API like
```rust
trait UserDefinedLogicalNode {
...
/// If only the specified indices of the output of node are needed
/// by the consumer, optionally returns a new UserDefinedNode that 
/// computes only those indices, in the specified order.
///
/// Return Ok(None), the default, if no such pushdown is possible
///
/// This is used to implement projection pushdown for user defined nodes
fn try_project(&self, indices: &[usize]) -> Result<Option<Arc<dyn Self>>> {
  Ok(None)
}
```

Reply:
> I suggest for this PR, only push down information about used expressions (don't try and incorporate a projection from the parent) and we can handle the parent projection information in a follow on PR.
This behavior may not be safe. Assuming a user defined operator similar to `LogicalPlan::Filter`. Following Plan
```
Projection(a,b,c)
--UserDefineFilter(d=0)
----TableScan(a,b,c,d)
```
If we don't incorporate requirements from the parent, user defined filter may insert `Projection(d)` below it. Which would produce following invalid plan:
```
Projection(a,b,c) (a,b,c is missing at its input)
--UserDefineFilter(d=0)
----Projection(d)
------TableScan(a,b,c,d)
```

Reply:
> I am thinking of an API like
> ```
> trait UserDefinedLogicalNode {
> ...
> /// If only the specified indices of the output of node are needed
> /// by the consumer, optionally returns a new UserDefinedNode that 
> /// computes only those indices, in the specified order.
> ///
> /// Return Ok(None), the default, if no such pushdown is possible
> ///
> /// This is used to implement projection pushdown for user defined nodes
> fn try_project(&self, indices: &[usize]) -> Result<Option<Arc<dyn Self>>> {
>   Ok(None)
> }
> ```

I think, implementers of the `try_project` needs to insert `LogicalPlan::Projection` on top of their input to pushdown projection. This might not be the case, if I am missing something. If that is the case, this might be cumbersome for the users. What about an API something like 
```
/// Returns the necessary expressions at each child to calculate argument.
/// assuming exprs refers to valid fields at the output of the operator.
/// Return Ok(None), the default, Cannot determine necessary expressions at the children to calculate given exprs.
fn necessary_children_exprs(&self, exprs: &[Expr]) -> Option<Vec<Vec<Expr>>> {
  Ok(None)
}
```
with an API like above, we can handle projection insertion outside the trait method.

Reply:
> Which would produce following invalid plan:

That is a good point. 

> What about an API something like

That is a great idea, though I don't fully understand why it uses `Expr`s 

How about a slight refinement in terms of input/output columns:

```rust
/// Returns the necessary input columns to this node required to compute 
/// the columns in the output schema
///
/// This is used for projection pushdown when DataFusion has determined that
/// only a subset of the output columns of this node are needed by its parents. 
/// This API is used to tell DataFusion which, if any, of the input columns are no longer
/// needed. 
/// 
/// Return `Ok(None)`, the default, if this information can not be determined
/// Returns `Ok(input_columns)` with the column indexes from this nodes' input that are
/// needed to compute `output_columns`)
fn necessary_children_exprs(&self, output_columns: &[usize]) -> Option<Vec<usize>> {
  Ok(None)
}
```

🤔

Reply:
This is much better. Knowing the mapping of the `Column`s is adequate. Using `Expr` has no additional support. Thanks @alamb for this suggestion. I will update this PR to use this mechanism. Then we can discuss further. I will let you know when it ready for further review.

Reply:
@alamb I introduced new API `fn necessary_children_exprs(&self, output_columns: &[usize]) -> Option<Vec<Vec<usize>>>` to this PR (where return type is changed to return result per child). This Pr is ready for further review.